File: examples/bvg.R
# Analysis of BMRestore samples from a bivariate Gaussian distribution
# with covariance matrix
#   1.2, 0.4
#   0.4, 0.8
# using a Gaussian regeneration distribution with identity covariance
# matrix
library(mvtnorm)
library(RColorBrewer)

# Target moments
mo1 <- c(0,0)
mo2 <- c(1.2, 0.8)

# True target covariance matrix
targ_cov <- matrix(c(1.2, 0.4, 0.4, 0.8), nrow=2)
targ_prec <- solve(targ_cov)

dtarg <- function(x){
  exp(-0.5 * as.vector(x %*% targ_prec %*% x))
}

# Log density of the target
ldtarg <- function(x){
  -0.5 * as.vector(x %*% targ_prec %*% x)
}

# Partial regeneration rate
kappa_partial <- function(x){
  0.5 * (sum((targ_prec %*% x)^2) - sum(diag(targ_prec)))
}

# Plot the target distribution, regeneration distribution and
# partial regeneration rate
n_grid <- 100
x_seq <- seq(from = -2.5, to = 2.5, length.out = n_grid)
y_seq <- x_seq
targ_dens <- matrix(0, nrow=n_grid, ncol=n_grid)
regen_dens <- matrix(0, nrow=n_grid, ncol=n_grid)
partial_regen_rate <- matrix(0, nrow=n_grid, ncol=n_grid)
for (i in 1:n_grid){
  for (j in 1:n_grid){
    xij <- c(x_seq[i],y_seq[j])
    targ_dens[i,j] <- dtarg(xij)
    regen_dens[i,j] <- dmvnorm(xij)
    partial_regen_rate[i,j] <- kappa_partial(xij)
  }
}

# Plot of the target distribution
contour(x_seq, y_seq, targ_dens, main='Target Distribution',
        xlab='x1', ylab='x2')

# Plot of the target and regeneration distributions
contour(x_seq, y_seq, targ_dens, main='Target and Regeneration Distributions',
        xlab='x1', ylab='x2')
contour(x_seq, y_seq, regen_dens, add=TRUE, col ='green')

# Plot of the target and regeneration distributions,
# as well as the partial regeneration rate
contour(x_seq, y_seq, targ_dens, xlab='x1', ylab='x2',
        main='Target and Regeneration Distributions, Partial Regeneration Rate')
contour(x_seq, y_seq, regen_dens, add=TRUE, col ='green')
contour(x_seq, y_seq, partial_regen_rate, add=TRUE, col = 'red')

# Partial regeneration rate at the origin
kappa_partial(c(0,0))

# Full regeneration rate
logC <- 2.07
kappa <- function(x){
  kappa_partial(x) + exp(logC + mvtnorm::dmvnorm(x, log=TRUE) - ldtarg(x))
}
optim(c(0,0), kappa)

# Plot contours
regen_rate <- matrix(0, nrow=n_grid, ncol=n_grid)
for (i in 1:n_grid){
  for (j in 1:n_grid){
    regen_rate[i,j] <- kappa(c(x_seq[i],y_seq[j]))
  }
}

# Full regeneration rate
contour(x_seq, y_seq, regen_rate, col = 'red', xlab='x1', ylab='x2',
        main='Full Regeneration Rate', levels=c(1, 4, 10, 20, 50, 100))

# Target distribution and full regeneration rate
contour(x_seq, y_seq, targ_dens, main='Target density and regeneration rate',
        xlab='x1', ylab='x2')
contour(x_seq, y_seq, regen_rate, col = 'red',
        levels=c(1, 4, 10, 20, 50, 100), add =TRUE)

#####################
# Short, detailed run
#####################

x <- read.table("bmrstr_x1.txt",
                quote="\"", comment.char="")
ts <- read.table("bmrstr_ts1.txt",
                quote="\"", comment.char="")
ts <- ts$V1
tours <- read.table("bmrstr_tours1.txt",
                quote="\"", comment.char="")
tours <- tours$V1

# 2d trace plots
ntours <- 10

plot(x[tours <= ntours,], type = 'n', xlab='X1', ylab='X2')
contour(x_seq, y_seq, targ_dens, col='gray', add=TRUE)
for (tour in 0:10){
  lines(x[tours==tour,])
}

# Same 2d plot, with colours
qual_col_pals <- brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector <- unlist(mapply(brewer.pal, qual_col_pals$maxcolors,
                            rownames(qual_col_pals)))
col_vector <- sample(col_vector)

plot(x[tours <= ntours,], type = 'n', xlab='X1', ylab='X2')
contour(x_seq, y_seq, targ_dens, col='gray', add=TRUE, nlevels = 5)
for (tour in 0:20){
  lines(x[tours==tour,], col=col_vector[tour], cex=2)
}

# Trace plots
for (i in 1:2){
  plot(ts, x[,i], type='n', xlab='t', ylab=paste0('X',i))
  for (tour in 0:(ntours-1)){
    lines(ts[tours==tour], x[tours==tour,i])
  }
}

# Trace plots: zoomed in
for (i in 1:2){
  plot(ts, x[,i], type='n', xlab='t', ylab=paste0('X',i),
       xlim=c(0,5))
  for (tour in 0:(ntours-1)){
    lines(ts[tours==tour], x[tours==tour,i])
  }
}

##############################
# Load samples from longer run
##############################
x2 <- read.table("bmrstr_x2.txt", quote="\"", comment.char="")
dim(x2)

# Marginals
for (i in 1:2){
  plot(density(x2[,i]), col='blue',
       main=paste0('Density Estimate of X', i))
  curve(dnorm(x, sd=sqrt(targ_cov[i,i])), add=TRUE)
}

# Analyse Monte Carlo samples
mo1_est <- colMeans(x2)
mo2_est <- colMeans(x2^2)

# RMSE
sqrt(mean((mo1 - mo1_est)^2))
sqrt(mean((mo2 - mo2_est)^2))

File: examples/bvg.cpp
/* Test class BMRestore
 *
 * Simulate from a bivariate Gaussian distribution with covariance matrix
 *      1.2, 0.4
 *      0.4, 0.8
 * using an isotropic Gaussian regeneration distribution.
 * Prints a short, detailed path to "bmrstr_mvg_x1.txt", "bmrstr_mvg_ts1.txt",
 * "bmrstr_mvg_tours1.txt" and a long path to "bmrstr_mvg_x2.txt".
 */

#include "bmrstr.h"
#include "log_post.h"
#include "mvg.h"
#include "regen_dist.h"
#include <armadillo>
#include <fstream>
#include <iostream>
#include <random>
#include <string>

#define LOGC 2.07
#define KAPPA_BAR 100.0
#define NTOURS1 100
#define OUTPUT_RATE1 1000
#define NTOURS2 100000
#define OUTPUT_RATE2 1.0

// Target log-density, gradient and laplacian
double ldtarg(const arma::vec &state, const arma::mat &precision);
void grad_ldtarg(const arma::vec &state,
                 arma::vec &grad,
                 const arma::mat &precision);
double lap_ldtarg(const arma::vec &state, const arma::mat &precision);

int main()
{
    // Dimension, constant C, kappa_bar
    int d = 2;
    
    // Target density
    // Covariance matrix
    arma::mat targ_cov({{1.2, 0.4},
                        {0.4, 0.8}});
    // Precision matrix
    arma::mat targ_prec = arma::inv_sympd(targ_cov);
    LogPost gauss(d, targ_prec, ldtarg, grad_ldtarg, lap_ldtarg);
    
    // Regeneration distribution has identity covariance matrix
    arma::mat redundant_mat(d, d, arma::fill::eye);
    RegenDist mu(d, redundant_mat, ld_mvg_iso, rmvg_iso);
    
    double logC = LOGC;
    double kappa_bar = KAPPA_BAR;
    int ntours = NTOURS1;
    double output_rate = OUTPUT_RATE1;

    // Short detailed run
    BMRestore X1(gauss, mu, logC, kappa_bar, ntours, output_rate);
    
    // Generate samples
    X1.gen_fixed_ntours();
    
    // Print to a file
    std::ofstream file;
    X1.print_output_states(file, "bmrstr_x1.txt");
    X1.print_output_times(file, "bmrstr_ts1.txt");
    X1.print_output_tour_number(file, "bmrstr_tours1.txt");
    
    // Long run
    ntours = NTOURS2;
    output_rate = OUTPUT_RATE2;
    
    BMRestore X2(gauss, mu, logC, kappa_bar, ntours, output_rate);
    
    X2.gen_fixed_ntours();
    
    // Print to file
    X2.print_output_states(file, "bmrstr_x2.txt");
    
    return 0;
}

double ldtarg(const arma::vec &state, const arma::mat &precision)
{
    arma::mat aux = state.t() * precision * state;
    return -0.5 * aux(0,0);
}

void grad_ldtarg(const arma::vec &state,
                 arma::vec &grad,
                 const arma::mat &precision)
{
    grad = -(precision * state);
}

double lap_ldtarg(const arma::vec &state, const arma::mat &precision)
{
    return -arma::trace(precision);
}
